src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+                plt.plot(x, y, 'b:')
+
+        corners = [workspace.scenicToSchematicCoords(corner) for corner in self.corners[:4]]
+        x, y = zip(*corners)
+        color = self.color if hasattr(self, 'color') else (1, 0, 0)
+        plt.fill(x, y, color=color)
+
+        frontMid = averageVectors(corners[0], corners[1])
+        baseTriangle = [frontMid, corners[2], corners[3]]
+        triangle = [averageVectors(p, spos, weight=0.5) for p in baseTriangle]
+        x, y = zip(*triangle)
+        plt.fill(x, y, "w")
+        plt.plot(x + (x[0],), y + (y[0],), color="k", linewidth=1)
+
+@distributionFunction
+def defaultSideSurface(occupiedSpace, dimension, positive, thresholds):
+    """Extracts a side surface from the occupiedSpace of an object.
+
+    This function is the default implementation for computing a region
+    representing a side surface of an object. This is done by keeping only the
+    faces of the object's ``occupiedSpace`` mesh that have normal 
+    vectors with a large/small enough x,y, or z component. For example, 
+    for the front surface of an object we would would keep all faces that 
+    had a normal vector with y component greater than ``thresholds[1][1]`` 
+    and for the back surface of an object we would keep all faces that
+    had a normal vector with y component less than ``thresholds[1][0]``.
+
+    Args:
+        occupiedSpace: The `occupiedSpace` region of the object to
+          extract the side surface from.
+        dimension: The target dimension who's component will be checked.
+        positive: If `False`, the target component must be less than 
+          the first value in the appropriate tuple. If `True`, the
+          component must be greater than the second value in the
+          appropriate tuple.
+        thresholds: A 3-tuple of 2-tuples, one for each dimension (x,y,z),
+          with each tuple containing the thresholds for a positive and
+          non-positive side in each dimension.
+        on_dimension: The on_dimension to be passed to the created surface.
+    """
+    # Extract mesh from object
+    obj_mesh = occupiedSpace.mesh.copy()
+
+    # Extract appropriate thresholds
+    threshold = thresholds[dimension][int(positive)]
+
+    # Drop all faces whose normal vector do not have a sufficiently
+    # large component.
+    face_normal_vals = obj_mesh.face_normals[:, dimension]
+    if positive:
+        face_mask = face_normal_vals >= threshold
+    else:
+        face_mask = face_normal_vals <= threshold
+
+    obj_mesh.faces = obj_mesh.faces[face_mask]
+    obj_mesh.remove_unreferenced_vertices()
+
+    # Check if the resulting surface is empty and return an appropriate region.
+    if not obj_mesh.is_empty:
+        return MeshSurfaceRegion(mesh=obj_mesh, center_mesh=False)
+    else:
+        return EmptyRegion(name="EmptyTopSurface")
+
+def canSee(position, orientation, visibleDistance, viewAngles, rayDensity, \
+           target, occludingObjects, debug=False):
```

Comment:
At 450+ lines this is a pretty mammoth function. It would help readability to factor most of the details out into separate functions, e.g. the front/back crossing checks, the computation of the angle ranges, the generation of the rays, etc. Then the structure of the function might become clear enough that you wouldn't need to repeat the description of the process in the docstring.

It might also be worth moving this into a new file (e.g. `visibility.py`) since it's unrelated to the rest of the file and is self-contained.

Reply:
Agreed! I've moved it into visibility.py and I've put the refactoring on the Todo list.